File: P0053_Max_Subarray_Sum.rs
// Problem Link: https://leetcode.com/problems/maximum-subarray/

use core::cmp::max;

pub fn max_sub_array(nums: Vec<i32>) -> i32 {
    return nums.iter()
               .scan(0, |acc, x| { *acc = max(*x, x + *acc); Some(*acc) })
               .max()
               .unwrap();
}
  

File: P0053_Max_Subarray_Sum.hs
-- Problem Link: https://leetcode.com/problems/maximum-subarray/

-- Solution 1
maxSubArray = maximum . scanl1 (\a b -> max b (a + b))

-- Solution #2 (APL inspired)
kite            = flip const
bfork l b r x y = b (l x y) (r x y) -- bfork = Binary Fork from APL (no bird combinator for it)

maxSubArray = maximum . scanl1 (bfork kite max (+))
